lib/puppet/pops/types/p_encrypted_type.rb

```diff
@@ -0,0 +1,354 @@
+      # No need to check for anything besides object identity
+      __id__ == o.__id__
+    end
+
+    # Returns identity equality since two Encrypted for the same value will always be different.
+    # @param o [Object] Any object to check for equality
+    # @return [Boolean]
+    # @api public
+    #
+    def ==(o)
+      self.eql?(o)
+    end
+
+    # Encrypts data using a certificate, a cipher name and Any data - returns an Encrypted.
+    # This is the API for creating an `Encrypted` value.
+    #
+    # Note that the `cipher` param accepts a single name, or an array of cipher names.
+    # The preferred cipher is returned from the array using the order in --accepted_ciphers if it is not empty.
+    # Otherwise, the single cipher, or the first given cipher in the given array is used.
+    #
+    # @param certificate [Puppet::SLL::Certificate] The certificate to use (i.e. for the receiver's public key)
+    # @param cipher [String, Array<String>] The name of the cipher to use - for example 'AES-256-CBC', or an array of names
+    # @param data [Object] Any puppet rich data data type
+    # @return [Puppet::Pops::Types::PEncryptedType::Encrypted] An Encrypted puppet data type value
+    # @api public
+    #
+    def self.encrypt(certificate, ciphers, data)
+
+      cipher = best_matching_cipher(ciphers)
+      if cipher.nil?
+        if ciphers.is_a?(Array)
+          raise ArgumentError, _("None of the cipher names \"%{ciphers}\" are supported. Supported ciphers: %{available_ciphers}") % {
+            available_ciphers: acceptable_and_available_ciphers,
+            ciphers: ciphers
+          }
+        else
+          raise ArgumentError, _("Unsupported cipher algorithm \"%{cipher_name}\". Supported ciphers: %{available_ciphers}") % {
+            available_ciphers: acceptable_and_available_ciphers,
+            cipher_name: cipher
+          }
+        end
+      end
+
+      key = certificate.content.public_key
+      fingerprint = certificate.fingerprint
+      data = serialize(data)
+      format = 'json,' + cipher
+
+      aes_encrypt = OpenSSL::Cipher.new(cipher).encrypt
+
+      # Use a random key
+      aes_encrypt.key = aes_key = aes_encrypt.random_key
+
+      # Use a random initialization vector (safer) - these 16 bytes are prepended to the clear text
+      # and dropped after decryption.
+      #
+      iv = aes_encrypt.random_iv
+
+      # Encrypt the data with this key
+      crypt = iv + aes_encrypt.update(data) << aes_encrypt.final
+      # Encrypt the random key with the public key
+      encrypted_key = rsa_key(key).public_encrypt(aes_key)
+
+      # Encrypt the fingerprint with public key
+      aes_encrypt.reset
+
+      # use different iv
+      iv = aes_encrypt.random_iv
+      encrypted_fingerprint = iv + aes_encrypt.update(fingerprint) << aes_encrypt.final
+
+      Encrypted.new(format, encrypted_key, crypt, encrypted_fingerprint)
```

Comment:
It's too bad ruby doesn't define methods for openssl's [`EVP_SealInit`](https://www.openssl.org/docs/man1.1.0/crypto/EVP_SealInit.html) which handles the "encrypting a symmetric session key with an asymmetric key" dance.